### PR TITLE
[dbus] add energy scan dbus api

### PR DESCRIPTION
**Pull request by Irving-cl**
This PR adds the 'energy scan' dbus API. This API is for the application level to choose a better channel.

### CODE REVIEW

File: src/utils/thread_helper.cpp [hunk 1]
```diff
@@ -131,6 +131,28 @@ void ThreadHelper::Scan(ScanHandler aHandler)
     }
 }
 
+void ThreadHelper::EnergyScan(EnergyScanHandler aHandler)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aHandler != nullptr);
```
Comment (jinran-google) [inline src/utils/thread_helper.cpp:138]: Return `BUSY` in the error case.

**Review by jinran-google — APPROVED** [review]
LGTM :+1: 

with a nit

**Review by xwang146 — APPROVED** [review]

File: src/dbus/server/introspect.xml [hunk 2]
```diff
@@ -27,6 +27,21 @@
...
+
+      The result struture definition is:
+      <literallayout>
+        struct {
+          uint8  channel
+          int8_t max_rssi
+        }
+      </literallayout>
+    -->
+    <method name="EnergyScan">
```
Comment (bukepo) [inline src/dbus/server/introspect.xml:41]: what channels to scan, preferred or  supported?

File: src/dbus/server/introspect.xml [hunk 3]
```diff
@@ -27,6 +27,21 @@
...
+      <arg name="result" type="a(yy)" direction="out"/>
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/dbus/server/introspect.xml:42]: Updated.

File: src/dbus/server/introspect.xml (hunk 2 above)
Comment (Irving-cl) [reply src/dbus/server/introspect.xml:43]: As discussed offline, we will always scan the preferred channels.

**Review by bukepo — APPROVED** [review]
LGTM:+1:

File: src/utils/thread_helper.hpp [hunk 4]
```diff
@@ -104,6 +105,15 @@ class ThreadHelper
      */
     void Scan(ScanHandler aHandler);
 
+    /**
+     * This method performs an IEEE 802.15.4 Energy Scan.
+     *
+     * @param[in] aScanDuration  The duration for the scan, in milliseconds.
+     * @param[in] aHandler  The scan result handler.
```
Comment (jwhui) [inline src/utils/thread_helper.hpp:112]:
```suggestion
     * @param[in] aScanDuration  The duration for the scan, in milliseconds.
     * @param[in] aHandler       The scan result handler.
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**